binary_search_tree/tree.py

```diff
@@ -14,35 +17,96 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(1)
+    # Space Complexity: O(N)
+    def add_helper(self,node,key,value):
+        if node is None:
+            return TreeNode(key,value)
+        if key < node.key:
+            node.left = self.add_helper(node.left, key, value)
+        else:
+            node.right = self.add_helper(node.right, key, value)
+
+        return node
+
+
     def add(self, key, value = None):
-        pass
+        self.root = self.add_helper(self.root,key,value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+        
+
+
+
+    # Time Complexity: O(N)
```

Comment:
For much of the same reason as I explained above in my comment about adding a node, finding a node takes O(log(n)) time assuming a balanced tree.

For example, with 7 items in a perfectly balanced tree, searching for one of the values in the nodes on the bottom will mean iterating 3 times. The exact number is actually ceiling(log_2(n)) + 1, where ceiling rounds up the fractional part to the next number, which keeps this at O(log(n)). (Not that we expected you to come up with this number, I just wanted to explain!)

Reply:
I guess as was said in Learn, unbalanced trees do trend to O(n), but I think we can assume balanced trees ideally.

Either way, not going to take points off.